napari/_vispy/layers/surface.py

```diff
@@ -42,51 +47,74 @@ def __init__(self, layer) -> None:
+                self.layer.texcoords,
+            )
+            self.node.attach(self._texture_filter)
+        elif self.layer._has_texture:
+            self._texture_filter.texture = self.layer.texture
+            self._texture_filter.texcoords = self.layer.texcoords
+
+        if self._texture_filter is not None:
+            self._texture_filter.enabled = self.layer._has_texture
+            self.node.update()
+
     def _on_data_change(self):
-        ndisplay = self.layer._slice_input.ndisplay
-        if len(self.layer._data_view) == 0 or len(self.layer._view_faces) == 0:
-            vertices = None
-            faces = None
-            vertex_values = np.array([0])
-        else:
+        vertices = None
+        faces = None
+        vertex_values = None
+        vertex_colors = None
+        if len(self.layer._data_view) and len(self.layer._view_faces):
             # Offsetting so pixels now centered
             # coerce to float to solve vispy/vispy#2007
             # reverse order to get zyx instead of xyz
             vertices = np.asarray(
                 self.layer._data_view[:, ::-1], dtype=np.float32
             )
-            # due to above xyz>zyx, also reverse order of faces to fix handedness of normals
+            # due to above xyz>zyx, also reverse order of faces to fix
+            # handedness of normals
             faces = self.layer._view_faces[:, ::-1]
-            vertex_values = self.layer._view_vertex_values
 
-        if vertices is not None and ndisplay == 3 and self.layer.ndim == 2:
-            vertices = np.pad(vertices, ((0, 0), (0, 1)))
+            values = self.layer._view_vertex_values
+            if len(values):
+                vertex_values = values
 
-        # manually detach filters when we go to 2D to avoid dimensionality issues
-        # see comments in napari#3475. The filter is set again after set_data!
-        if ndisplay == 2:
-            filt = self.node.shading_filter
-            try:
-                self.node.detach(filt)
-                self.node.shading = None
-                self.node.shading_filter = None
-            except ValueError:
-                # sometimes we try to detach non-attached filters, which causes a ValueError
-                pass
+            colors = self.layer._view_vertex_colors
+            if len(colors):
+                vertex_colors = colors
+
+        # making sure the vertex data is 3D prevents shape errors with
+        # attached filters, instead of trying to attach/detach each time
+        if vertices is not None and vertices.shape[-1] == 2:
+            vertices = np.pad(vertices, ((0, 0), (0, 1)))
+        assert vertices is None or vertices.shape[-1] == 3
 
         self.node.set_data(
-            vertices=vertices, faces=faces, vertex_values=vertex_values
+            vertices=vertices,
+            faces=faces,
+            vertex_values=vertex_values,
+            vertex_colors=vertex_colors,
+            color="white",
```

Comment:
can you explain this combination to me? ie why is there both `vertex_colors=` and `color=`? Ideally I think adding a comment here will be useful as generally, hardcoded values are a "smell" — it's good to reassure readers that the smell is totally fine and expected and they are probably not going to die from it 😂

Reply:
We might be able to remove this instead. It's only used as the base mesh color if all the other coloring options are `None` but I think `vertex_values` defaults to all ones even if not provided. Leaving it unset otherwise can make transparent textures look off because the default is not white.

Reply:
all of that is useful info that could go in the comment! =) but e.g. you could go DEFAULT_TEXTURE_COLOR="white", and then, yeah, as you say, `if vertex_colors is None: kwargs += {'color': DEFAULT_TEXTURE_COLOR}` or something like that?

Reply:
I've just removed this for now. The mesh will always have a color because `vertex_values` is always set (defaulting to ones, which is white with the default colormap).